### DISCUSSION TITLE
How to differentiate unique constraint errors?

### DISCUSSION THREAD

**Discussion by ric980** [jackc/pgx#1616] [General]
Hello,

I have a unique constraint on an email column. When a user is created with already existing email, I want to return 409 from my API. Now I return 500 because I cannot differentiate between a unique constraint error and normal error. I don't want to make another call to the database just to check if a user with that email already exists. I don't want to parse the error as string and pattern match for 'duplicate key'. How to properly handle this case with the pgx driver?

Thanks.

**Comment by jackc** [comment]
See https://github.com/jackc/pgx/wiki/Error-Handling and https://pkg.go.dev/github.com/jackc/pgx/v5@v5.3.1/pgconn#PgError for how to get the actual error fields. Then https://github.com/jackc/pgerrcode may be of use for named errors that map to SQLSTATE error codes (e.g. 23503 - foreign key violation).

**Comment by ric980** [comment]
Thank you. The links are very helpful indeed. The code I was looking for was 23505.
